proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+of scope.
+
+
+Implicit lifting and deriving ``Lift`` instances
+################################################
+.. _lift_instances:
+
+``Lift`` instances are used to provide serialisation-based cross-stage
+persistence.  For example, a typical ``Lift`` instance looks like::
+
+    data MInt = Some Int | None
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+The presence of this instance means the following declaration will be accepted::
+
+  foo :: MInt -> Q Exp
+  foo x = [| x |]  -- implicitly becomes [| $(lift x) |]
+
+Defining a ``Lift`` instance requires the datatype constructors to be available
+both at compile-time and runtime, so defining ``Lift`` within the same module as
+the datatype itself requires path-based cross-stage persistence.  Operationally,
+``None`` and ``Some`` are needed both at compile-time *and*  runtime since they
+are both matched on at compile time, and also persisted to be spliced in the
+future into a program that can make use of them at runtime. As a result, it
+isn't possible to define or derive a (non-orphan) ``Lift`` instance under
+``NoImplicitStagePersistence``.
+
+An orphan ``Lift`` instance can be defined thus::
+
+  module M where
+    data MInt = Some Int | None
+
+  module N where
+    import M
+    import quote M
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+This isn't technically problematic, rather it is just a result of what ``Lift``
+means. However, it means some users may need to modify their use of ``Lift``
+instances if they wish to benefit more from ``NoImplicitStagePersistence``.
+Users are free to use ``ImplicitStagePersistence`` in selected modules to allow
+defining ``Lift`` instances, but doing so means all the dependencies of the
+module will need to be available both at compile-time and runtime.
+
+As a general rule, ``Lift`` instances should be defined only for simple
+datatypes near the root of the module hierarchy of an application.
+
+Just as ``NoImplicitStagePersistence`` allows users to disable implicit
+path-based cross-stage persistence, it would make sense to have an extension
+flag to disable implicit lifting (serialisation-based persistence).  This would
+allow the programmer to ensure they are explicit about where calls to ``lift``
+occur in their programs, which is sometimes desirable when using staging for
+runtime performance.  We intend to bring forward a separate proposal for this,
+as it is otherwise orthogonal to the current proposal.
+
+
+Implicit ``Prelude`` imports
+############################
+
+``Prelude`` does not get implicitly imported with ``splice`` or ``quote``. Therefore
+if you wish to use definitions from your ``Prelude`` module at non-zero levels
+then you have to explicitly import it at that level.
```

Comment:
It's worth discussing the alternative that `Prelude` is imported at all 3 levels. I don't see the downsides, yet. It doesn't look innocent, so maybe it should be made explicit in the _Specification_ section.

Reply:
If you require `Prelude` at all 3 levels implicitly, then how do you remove one of these dependencies if you wish to use a different `Prelude` at one particular level? 

Is your proposal in that case you would have to use `NoImplicitPrelude` to turn off all 3 imports?

It seemed to me undesirable to automatically add these additional imports and hence dependencies on certain stages unless they were actually used.

Reply:
It matters for building `base` at least!

Reply:
> Is your proposal in that case you would have to use `NoImplicitPrelude` to turn off all 3 imports?

That would be my intuition, yes. It's fine if you don't like this alternative, but it should be features in the _Alternatives_ section together with your arguments for choosing your design over this one.

Reply:
@Ericson2314 
> It matters for building `base` at least!

Could you elaborate?

Reply:
I think @Ericson2314 is pointing out that when you are building `base` then an implicit `import splice` would add a dependency on `Prelude` that you might not want.

Perhaps that doesn't matter so much for `base` as there isn't any `TemplateHaskell` used there. But it could be a problem if you are defining your own `base-like` library which defines its own `Prelude` module, at which point you would lose some control. If you are doing this though, you probably have got `NoImplicitPrelude` turned on.. so I'm not sure it matters much either way.

Reply:
Yeah I don't really like saying "Prelude as imported at 1 level or more levels based on whether `TemplateHaskell` is in use". IMO it should be true that with `NoImplicitStagePersistence` on, the presence or absence of `TemplateHaskell` should not affect imports, implicit or explicit. That means it shouldn't affect `Prelude`.